### PR TITLE
Turbulence straight channel case 

### PR DESCRIPTION
**Pull request by FengWang3119**
Turbulence straight channel case, periodic inlet and outlet, driven by gravity force. 
The name of the case file is 2d_turbulent_channel_PBC 
The turbulence model is in the folder "for_2D_build", named k-epsilon_turbulent_model.
Please check the code.

### CODE REVIEW

File: tests/user_examples/extra_src/for_2D_build/k-epsilon_turbulent_model.h [hunk 1]
```diff
@@ -0,0 +1,458 @@
...
+		Real turbu_const_E;
+		Real C_mu;
+		Real TurbulentIntensity;
+		//** Closure coefficients for K *
+		Real sigma_k;
+
+		//** Closure coefficients for Epsilon *
+		Real C_l, C_2;
+		Real sigma_E;
+	};
```
Comment (ChiZhangatTUM) [inline tests/user_examples/extra_src/for_2D_build/k-epsilon_turbulent_model.h:57]: For variable member, we following the naming patter "variable_name_" , all the in lowercase and ended with under line "_"

File: tests/user_examples/extra_src/for_2D_build/k-epsilon_turbulent_model.h [hunk 2]
```diff
@@ -0,0 +1,458 @@
...
+//=================================================================================================//
```
Comment (ChiZhangatTUM) [inline tests/user_examples/extra_src/for_2D_build/k-epsilon_turbulent_model.h:58]: remove this line in header file.

File: tests/user_examples/extra_src/for_2D_build/k-epsilon_turbulent_model.h [hunk 3]
```diff
@@ -0,0 +1,458 @@
...
+	template <typename... T>
+	class BaseTurtbulentModel;
+	
+	template <class DataDelegationType>
+	class BaseTurtbulentModel<Base, DataDelegationType>
```
Comment (ChiZhangatTUM) [inline tests/user_examples/extra_src/for_2D_build/k-epsilon_turbulent_model.h:63]: "Base" is not a good name pattern.

File: tests/user_examples/extra_src/for_2D_build/k-epsilon_turbulent_model.h [hunk 4]
```diff
@@ -0,0 +1,458 @@
...
+		Real sigma_k;
+
+		//** Closure coefficients for Epsilon *
+		Real C_l, C_2;
+		Real sigma_E;
+	};
+//=================================================================================================//
+	template <typename... T>
+	class BaseTurtbulentModel;
+
```
Comment (ChiZhangatTUM) [inline tests/user_examples/extra_src/for_2D_build/k-epsilon_turbulent_model.h:61]: Why we need these two line?

File: tests/user_examples/extra_src/for_2D_build/k-epsilon_turbulent_model.h [hunk 5]
```diff
@@ -0,0 +1,458 @@
...
+	template <>
+	class GetVelocityGradient<Inner<>> : public GetVelocityGradient<Base, FluidDataInner>
+	{
+	public:
+		explicit GetVelocityGradient(BaseInnerRelation& inner_relation)
+			: GetVelocityGradient<Base, FluidDataInner>(inner_relation) {};
+		virtual ~GetVelocityGradient() {};
+		void interaction(size_t index_i, Real dt = 0.0);
+	};
+	using GetVelocityGradientInner = GetVelocityGradient<Inner<>>;
```
Comment (ChiZhangatTUM) [inline tests/user_examples/extra_src/for_2D_build/k-epsilon_turbulent_model.h:111]: Is possible to write the gradient in a general way? For example, WeakGradientComputation & StrongGradientComputatoin drived from BaseGraidentComputation and velocity gradient is computed as 
StrongGradientComputation<Vecd> velocity_gradient("Velocity", complex_contact)

File: tests/user_examples/extra_src/for_2D_build/k-epsilon_turbulent_model.h [hunk 6]
```diff
@@ -0,0 +1,458 @@
...
+		template <class BaseRelationType>
+		explicit GetVelocityGradient(BaseRelationType& base_relation);
+		virtual ~GetVelocityGradient() {};
+	protected:
+		StdLargeVec<Vecd>& vel_, & pos_;
+		StdLargeVec<Matd>& velocity_gradient_;
+		StdLargeVec<int>& is_near_wall_P1_; //** This is used to specially treat near wall region  *
+
+		//**For test*
+		StdLargeVec<Matd> velocity_gradient_wall;
```
Comment (ChiZhangatTUM) [inline tests/user_examples/extra_src/for_2D_build/k-epsilon_turbulent_model.h:99]: Again, variable member ended with under line.

File: tests/user_examples/extra_src/for_2D_build/k-epsilon_turbulent_model.h [hunk 7]
```diff
@@ -0,0 +1,458 @@
...
+	};
+	//** Inner part *
+	template <>
+	class GetVelocityGradient<Inner<>> : public GetVelocityGradient<Base, FluidDataInner>
+	{
```
Comment (ChiZhangatTUM) [inline tests/user_examples/extra_src/for_2D_build/k-epsilon_turbulent_model.h:104]: "Inner" and "Base" has no specific meaning.

File: tests/user_examples/extra_src/for_2D_build/k-epsilon_turbulent_model.h (hunk 4 above)
Comment (FengWang3119) [reply tests/user_examples/extra_src/for_2D_build/k-epsilon_turbulent_model.h:61]: Just to sperate different parts, it can be moved.

**Comment by FengWang3119** [comment]
Thanks! I will modify the code according to your comments.

File: tests/user_examples/extra_src/for_2D_build/k-epsilon_turbulent_model.h (hunk 5 above)
Comment (FengWang3119) [reply tests/user_examples/extra_src/for_2D_build/k-epsilon_turbulent_model.h:111]: OK, I will try to write a common one,

**Review by Xiangyu-Hu — APPROVED** [review]

**Pull request merged**